File: topics/openacc/practicals/basics/dot_openacc.cpp
#include <iostream>

#define NO_CUDA
#include "util.h"

// host implementation of dot product
double dot_host(const double *x, const double *y, int n) {
    double sum = 0;

    #pragma omp parallel for reduction(+:sum)
    for (int i = 0; i < n; ++i) {
        sum += x[i]*y[i];
    }

    return sum;
}

double dot_gpu(const double *x, const double *y, int n) {
    double sum = 0;
    int i;

    // TODO: Offload this loop to the GPU
    for (i = 0; i < n; ++i) {
        sum += x[i]*y[i];
    }

    return sum;
}

int main(int argc, char **argv) {
    size_t pow  = read_arg(argc, argv, 1, 2);
    size_t n = 1 << pow;

    auto size_in_bytes = n * sizeof(double);

    std::cout << "dot product OpenACC of length n = " << n
              << " : " << size_in_bytes/(1024.*1024.) << "MB\n";

    auto x_h = malloc_host<double>(n, 2.);
    auto y_h = malloc_host<double>(n);
    for(auto i = 0; i < n; ++i) {
        y_h[i] = rand() % 10;
    }

    auto time_host = get_time();
    auto expected = dot_host(x_h, y_h, n);
    time_host = get_time() - time_host;

    auto time_gpu = get_time();
    auto result = dot_gpu(x_h, y_h, n);
    time_gpu = get_time() - time_gpu;
    std::cout << "expected " << expected << " got " << result << ": "
              << (std::abs(expected - result) < 1.e-6 ? "success\n" : "failure\n");
    std::cout << "Host kernel took " << time_host << " s\n";
    std::cout << "GPU kernel took " << time_gpu << " s\n";
    return 0;
}

File: topics/openacc/practicals/basics/dot_openacc.f90
subroutine dotprod_gpu(n, x, y, res)
  integer, intent(in) :: n
  real(kind(0d0)), intent(in) :: x(n)
  real(kind(0d0)), intent(in) :: y(n)
  real(kind(0d0)), intent(out) :: res

  res = 0
  ! TODO: Offload this loop to the GPU
  do i = 1,n
     res = res + x(i)*y(i)
  enddo

end subroutine dotprod_gpu

subroutine dotprod_host(n, x, y, res)
  integer, intent(in) :: n
  real(kind(0d0)), intent(in) :: x(n)
  real(kind(0d0)), intent(in) :: y(n)
  real(kind(0d0)), intent(out) :: res

  res = 0
  !$omp parallel do reduction(+:res)
  do i = 1,n
     res = res + x(i)*y(i)
  enddo

end subroutine dotprod_host

program main
  use util
  implicit none

  integer n, pow, err
  real(kind(0d0)), dimension(:), allocatable :: x, y
  real(kind(0d0)) :: result, expected
  real(kind(0d0)) :: time_gpu, time_host

  pow = read_arg(1, 2)
  n = 2**pow

  write(*, '(a i0)') 'dot product of length n = ', n
  allocate(x(n), y(n), stat=err)
  if (err /= 0) then
     stop 'failed to allocate arrays'
  endif

  call random_seed()

  x(:) = 2.0d0
  call random_number(y)
  y(:) = 10*y(:)

  time_host = get_time()
  call dotprod_host(n, x, y, expected)
  time_host = get_time() - time_host

  time_gpu = get_time()
  call dotprod_gpu(n, x, y, result)
  time_gpu = get_time() - time_gpu

  if (abs(expected - result) > 1.e-6) then
     write(*, '(a e0.6 a e0.6 a)') 'expected ', expected, ' got, ', result, &
          ': failure'
  else
     write(*, '(a e0.6 a e0.6 a)') 'expected ', expected, ' got, ', result, &
          ': success'
  endif

end program main
